File: app/src/main/java/example/com/crackle/model/Movie.java
package example.com.crackle.model;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

@Entity(tableName = "movie")
public class Movie implements Parcelable {

    @PrimaryKey
    @ColumnInfo(name = "movie_id")
    @SerializedName("id")
    private int movieId;

    @ColumnInfo(name = "poster_path")
    @SerializedName("poster_path")
    private String imageUrl;

    @ColumnInfo(name = "backdrop_path")
    @SerializedName("backdrop_path")
    private String backdropImageUrl;

    @ColumnInfo(name = "title")
    @SerializedName("title")
    private String title;

    @ColumnInfo(name = "overview")
    @SerializedName("overview")
    private String plot;

    @ColumnInfo(name = "popularity")
    @SerializedName("popularity")
    private double popularity;

    @ColumnInfo(name = "vote_average")
    @SerializedName("vote_average")
    private double userRating;

    @ColumnInfo(name = "release_date")
    @SerializedName("release_date")
    private String releaseDate;

    @ColumnInfo(name = "original_language")
    @SerializedName("original_language")
    private String language;

    @ColumnInfo(name = "runtime")
    @SerializedName("runtime")
    private int duration;

    @ColumnInfo(name = "original_title")
    @SerializedName("original_title")
    private String originalTitle;

    @ColumnInfo(name = "homepage")
    @SerializedName("homepage")
    private String homepage;

    @Ignore
    @SerializedName("genre_ids")
    private ArrayList<Integer> genres;

    @Ignore
    @SerializedName("images")
    private ImageResults imageResults;

    @Ignore
    @SerializedName("videos")
    private VideoResults videoResults;

    @Ignore
    @SerializedName("releases")
    private CertificationResults certificationResults;

    @ColumnInfo(name = "is_favorite")
    private boolean isFavorite;

    @Ignore
    public Movie() {
    }

    //this constructor will be used by Room
    public Movie(int movieId, String imageUrl, String backdropImageUrl, String title, String plot, double popularity, double userRating, String releaseDate, String language, int duration, String originalTitle, String homepage, boolean isFavorite) {
        this.movieId = movieId;
        this.imageUrl = imageUrl;
        this.backdropImageUrl = backdropImageUrl;
        this.title = title;
        this.plot = plot;
        this.popularity = popularity;
        this.userRating = userRating;
        this.releaseDate = releaseDate;
        this.language = language;
        this.duration = duration;
        this.originalTitle = originalTitle;
        this.homepage = homepage;
        this.isFavorite = isFavorite;
    }

    @Ignore
    public Movie(int movieId, String imageUrl, String backdropImageUrl, String title, String plot, double popularity, double userRating, String releaseDate, String language, int duration, String originalTitle, String homepage, ArrayList<Integer> genres, ImageResults imageResults, VideoResults videoResults, CertificationResults certificationResults, boolean isFavorite) {
        this.movieId = movieId;
        this.imageUrl = imageUrl;
        this.backdropImageUrl = backdropImageUrl;
        this.title = title;
        this.plot = plot;
        this.popularity = popularity;
        this.userRating = userRating;
        this.releaseDate = releaseDate;
        this.language = language;
        this.duration = duration;
        this.originalTitle = originalTitle;
        this.homepage = homepage;
        this.genres = genres;
        this.imageResults = imageResults;
        this.videoResults = videoResults;
        this.certificationResults = certificationResults;
        this.isFavorite = isFavorite;
    }


    @Ignore
    public Movie(Parcel source) {
        movieId = source.readInt();
        imageUrl = source.readString();
        backdropImageUrl = source.readString();
        title = source.readString();
        plot = source.readString();
        popularity = source.readDouble();
        userRating = source.readDouble();
        releaseDate = source.readString();
        language = source.readString();
        duration = source.readInt();
        originalTitle = source.readString();
        homepage = source.readString();
        genres = (ArrayList<Integer>) source.readSerializable();
        imageResults = source.readParcelable(ImageResults.class.getClassLoader());
        videoResults = source.readParcelable(VideoResults.class.getClassLoader());
        certificationResults = source.readParcelable(CertificationResults.class.getClassLoader());
        isFavorite = source.readByte() != 0;
    }

    public int getMovieId() {
        return movieId;
    }

    public void setMovieId(int movieId) {
        this.movieId = movieId;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getBackdropImageUrl() {
        return backdropImageUrl;
    }

    public void setBackdropImageUrl(String backdropImageUrl) {
        this.backdropImageUrl = backdropImageUrl;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getPlot() {
        return plot;
    }

    public void setPlot(String plot) {
        this.plot = plot;
    }

    public double getPopularity() {
        return popularity;
    }

    public void setPopularity(double popularity) {
        this.popularity = popularity;
    }

    public double getUserRating() {
        return userRating;
    }

    public void setUserRating(double userRating) {
        this.userRating = userRating;
    }

    public String getReleaseDate() {
        return releaseDate;
    }

    public void setReleaseDate(String releaseDate) {
        this.releaseDate = releaseDate;
    }

    public String getLanguage() {
        return language;
    }

    public void setLanguage(String language) {
        this.language = language;
    }

    public int getDuration() {
        return duration;
    }

    public void setDuration(int duration) {
        this.duration = duration;
    }

    public String getOriginalTitle() {
        return originalTitle;
    }

    public void setOriginalTitle(String originalTitle) {
        this.originalTitle = originalTitle;
    }

    public String getHomepage() {
        return homepage;
    }

    public void setHomepage(String homepage) {
        this.homepage = homepage;
    }

    public ArrayList<Integer> getGenres() {
        return genres;
    }

    public void setGenres(ArrayList<Integer> genres) {
        this.genres = genres;
    }

    public ImageResults getImageResults() {
        return imageResults;
    }

    public void setImageResults(ImageResults imageResults) {
        this.imageResults = imageResults;
    }

    public VideoResults getVideoResults() {
        return videoResults;
    }

    public void setVideoResults(VideoResults videoResults) {
        this.videoResults = videoResults;
    }

    public CertificationResults getCertificationResults() {
        return certificationResults;
    }

    public void setCertificationResults(CertificationResults certificationResults) {
        this.certificationResults = certificationResults;
    }

    public boolean isFavorite() {
        return isFavorite;
    }

    public void setFavorite(boolean favorite) {
        isFavorite = favorite;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(movieId);
        dest.writeString(imageUrl);
        dest.writeString(backdropImageUrl);
        dest.writeString(title);
        dest.writeString(plot);
        dest.writeDouble(popularity);
        dest.writeDouble(userRating);
        dest.writeString(releaseDate);
        dest.writeString(language);
        dest.writeInt(duration);
        dest.writeString(originalTitle);
        dest.writeString(homepage);
        dest.writeSerializable(genres);
        dest.writeParcelable(imageResults, 0);
        dest.writeParcelable(videoResults, 0);
        dest.writeParcelable(certificationResults, 0);
        dest.writeByte((byte) (isFavorite ? 1 : 0));
    }

    public static final Parcelable.Creator<Movie> CREATOR = new Parcelable.Creator<Movie>() {
        @Override
        public Movie createFromParcel(Parcel source) {
            return new Movie(source);
        }

        @Override
        public Movie[] newArray(int size) {
            return new Movie[size];
        }
    };
}

File: app/src/main/java/example/com/crackle/model/Movie.kt
package example.com.crackle.model

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName

@Entity(tableName = "movie")
class Movie : Parcelable {
    @PrimaryKey
    @ColumnInfo(name = "movie_id")
    @SerializedName("id")
    var movieId = 0
    @ColumnInfo(name = "poster_path")
    @SerializedName("poster_path")
    var imageUrl: String? = null
    @ColumnInfo(name = "backdrop_path")
    @SerializedName("backdrop_path")
    var backdropImageUrl: String? = null
    @ColumnInfo(name = "title")
    @SerializedName("title")
    var title: String? = null
    @ColumnInfo(name = "overview")
    @SerializedName("overview")
    var plot: String? = null
    @ColumnInfo(name = "popularity")
    @SerializedName("popularity")
    var popularity = 0.0
    @ColumnInfo(name = "vote_average")
    @SerializedName("vote_average")
    var userRating = 0.0
    @ColumnInfo(name = "release_date")
    @SerializedName("release_date")
    var releaseDate: String? = null
    @ColumnInfo(name = "original_language")
    @SerializedName("original_language")
    var language: String? = null
    @ColumnInfo(name = "runtime")
    @SerializedName("runtime")
    var duration = 0
    @ColumnInfo(name = "original_title")
    @SerializedName("original_title")
    var originalTitle: String? = null
    @ColumnInfo(name = "homepage")
    @SerializedName("homepage")
    var homepage: String? = null
    @SerializedName("genre_ids")
    var genres: ArrayList<Int>? = null
    @Ignore
    @SerializedName("images")
    var imageResults: ImageResults? = null
    @Ignore
    @SerializedName("videos")
    var videoResults: VideoResults? = null
    @Ignore
    @SerializedName("releases")
    var certificationResults: CertificationResults? = null
    @ColumnInfo(name = "is_favorite")
    var isFavorite = false

    @Ignore
    constructor()

    //this constructor will be used by Room
    constructor(movieId: Int, imageUrl: String?, backdropImageUrl: String?, title: String?, plot: String?, popularity: Double, userRating: Double, releaseDate: String?, language: String?, duration: Int, originalTitle: String?, homepage: String?, isFavorite: Boolean) {
        this.movieId = movieId
        this.imageUrl = imageUrl
        this.backdropImageUrl = backdropImageUrl
        this.title = title
        this.plot = plot
        this.popularity = popularity
        this.userRating = userRating
        this.releaseDate = releaseDate
        this.language = language
        this.duration = duration
        this.originalTitle = originalTitle
        this.homepage = homepage
        this.isFavorite = isFavorite
    }

    @Ignore
    constructor(movieId: Int, imageUrl: String?, backdropImageUrl: String?, title: String?, plot: String?, popularity: Double, userRating: Double, releaseDate: String?, language: String?, duration: Int, originalTitle: String?, homepage: String?, genres: ArrayList<Int>?, imageResults: ImageResults?, videoResults: VideoResults?, certificationResults: CertificationResults?, isFavorite: Boolean) {
        this.movieId = movieId
        this.imageUrl = imageUrl
        this.backdropImageUrl = backdropImageUrl
        this.title = title
        this.plot = plot
        this.popularity = popularity
        this.userRating = userRating
        this.releaseDate = releaseDate
        this.language = language
        this.duration = duration
        this.originalTitle = originalTitle
        this.homepage = homepage
        this.genres = genres
        this.imageResults = imageResults
        this.videoResults = videoResults
        this.certificationResults = certificationResults
        this.isFavorite = isFavorite
    }

    @Ignore
    constructor(source: Parcel) {
        movieId = source.readInt()
        imageUrl = source.readString()
        backdropImageUrl = source.readString()
        title = source.readString()
        plot = source.readString()
        popularity = source.readDouble()
        userRating = source.readDouble()
        releaseDate = source.readString()
        language = source.readString()
        duration = source.readInt()
        originalTitle = source.readString()
        homepage = source.readString()
        genres = source.readSerializable() as ArrayList<Int>?
        imageResults = source.readParcelable(ImageResults::class.java.classLoader)
        videoResults = source.readParcelable(VideoResults::class.java.classLoader)
        certificationResults = source.readParcelable(CertificationResults::class.java.classLoader)
        isFavorite = source.readByte().toInt() != 0
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(movieId)
        dest.writeString(imageUrl)
        dest.writeString(backdropImageUrl)
        dest.writeString(title)
        dest.writeString(plot)
        dest.writeDouble(popularity)
        dest.writeDouble(userRating)
        dest.writeString(releaseDate)
        dest.writeString(language)
        dest.writeInt(duration)
        dest.writeString(originalTitle)
        dest.writeString(homepage)
        dest.writeSerializable(genres)
        dest.writeParcelable(imageResults, 0)
        dest.writeParcelable(videoResults, 0)
        dest.writeParcelable(certificationResults, 0)
        dest.writeByte((if (isFavorite) 1 else 0).toByte())
    }

    companion object {
        @JvmField val CREATOR: Creator<Movie?> =  object : Creator<Movie?> {
            override fun createFromParcel(source: Parcel): Movie? {
                return Movie(source)
            }

            override fun newArray(size: Int): Array<Movie?> {
                return arrayOfNulls(size)
            }
        }
    }
}